### ISSUE TITLE
关于Prompt_Tokens的问题

### ISSUE THREAD

**Issue by dong-0412** [sagizty/VPT#5]
if self.VPT_type == "Deep":

            Prompt_Token_num = self.Prompt_Tokens.shape[1]

            for i in range(len(self.blocks)):
                # concatenate Prompt_Tokens
                Prompt_Tokens = self.Prompt_Tokens[i].unsqueeze(0)
                # firstly concatenate
                x = torch.cat((x, Prompt_Tokens.expand(x.shape[0], -1, -1)), dim=1)
                num_tokens = x.shape[1]
                # lastly remove, a genius trick
                x = self.blocks[i](x)[:, :num_tokens - Prompt_Token_num]

        else:  # self.VPT_type == "Shallow"
            Prompt_Token_num = self.Prompt_Tokens.shape[1]

            # concatenate Prompt_Tokens
            Prompt_Tokens = self.Prompt_Tokens.expand(x.shape[0], -1, -1)
            x = torch.cat((x, Prompt_Tokens), dim=1)
            num_tokens = x.shape[1]
            # Sequntially procees
            x = self.blocks(x)[:, :num_tokens - Prompt_Token_num]

你在代码里写的是把prompt直接cat到cls+x的后面，然后让这个cat后的整体经过一个Transformer Encoder Layer，算出的结果直接去掉最后的Prompt_Token_num个，我的疑问就来自这里，我的印象里Transformer Encoder Layer的输出和输入不是一对一的而是每个输出都关注全部的输入信息，那这样简单的去掉后面的Prompt_Token_num个会不会有信息的丢失之类的问题出现。

希望您抽空解决我困惑

**Comment by sagizty** [comment]
Thanks for your question, they’re very nice and they right hit the point. Allow me to explain further:
1. ViT and other encoders are token-wise projection, the prompt tokens therefore influence the other tokens, just as you said.
2. To further explore this, you can try concatenate different amounts of tokens. In this case, although the model structure is fixed(and model parameters are frozen), the model still work, and the result are influenced.
3. Please notice: I am not the author of VPT, therefore I am only answering the enquires based on my understanding.
4. It’s better to put your questions in English so everyone can enjoy.
